sdk/search/Azure.Search.Documents/src/Batching/Publisher{T}.cs

```diff
@@ -0,0 +1,505 @@
+        /// <returns>Task for flushing.</returns>
+        protected virtual async Task OnFlushedAsync(CancellationToken cancellationToken)
+        {
+            await PublishAsync(cancellationToken).ConfigureAwait(false);
+
+            // Wake up anyone who was blocked on a flush finishing
+            TaskCompletionSource<object> previous = _flushCompletionSource;
+            _flushCompletionSource = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
+            previous.SetResult(null);
+        }
+        #endregion Message Consumer
+
+        #region Dispose
+        /// <summary>
+        /// Clean up any resources.
+        /// </summary>
+        async ValueTask IAsyncDisposable.DisposeAsync() =>
+            await DisposeAsync(async: true).ConfigureAwait(false);
+
+        /// <summary>
+        /// Clean up any resources.
+        /// </summary>
+        void IDisposable.Dispose() =>
+            DisposeAsync(async: false).EnsureCompleted();
+
+        /// <summary>
+        /// Dispose the sender and flush any remaining indexing actions that
+        /// haven't been sent yet.  This will block until everything's been
+        /// sent.
+        /// </summary>
+        /// <param name="async">Whether to call this sync or async.</param>
+        /// <returns>A Task that will wait until we're disposed.</returns>
+        public async Task DisposeAsync(bool async)
+        {
+            if (Interlocked.CompareExchange(ref _disposed, 1, 0) == 0)
+            {
+                StopTimer();
+                await FlushInternal(async, PublisherCancellationToken).ConfigureAwait(false);
+                _channel?.Writer.TryComplete();
+                if (async)
+                {
+                    await _readerLoop.ConfigureAwait(false);
+                }
+                else
+                {
+                    #pragma warning disable AZC0104 // Use EnsureCompleted() directly on asynchronous method return value.
+                    _readerLoop.EnsureCompleted();
+                    #pragma warning restore AZC0104 // Use EnsureCompleted() directly on asynchronous method return value.
+                }
+                _flushCompletionSource.SetCanceled();
+            }
+        }
+
+        /// <summary>
+        /// Ensure nobody's trying to use this after it's been disposed.
+        /// </summary>
+        private void EnsureNotDisposed()
+        {
+            if (_disposed > 0)
+            {
+                throw new ObjectDisposedException(nameof(SearchIndexingBufferedSender<T>));
+            }
+        }
+        #endregion Dispose
+
+        #region Publishing
+        /// <summary>
+        /// Check if we have a full batch ready to send.
+        /// </summary>
+        /// <returns>If we have a full batch ready to send.</returns>
+        private bool HasFullBatch() => IndexingActionsCount >= BatchActionSize;
```

Comment:
nit: should this be a property?

Reply:
I hope to make this 10x more complex in the near future so I went with a method for now.